sw/device/lib/dif/dif_rstmgr.c

```diff
@@ -0,0 +1,215 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_rstmgr.h"
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+#include "rstmgr_regs.h"  // Generated.
+
+_Static_assert(kDifRstmgrCausePowerOn == RSTMGR_RESET_INFO_CAUSE_POR,
+               "kDifRstmgrCausePowerOn must match the register definition!");
+_Static_assert(
+    kDifRstmgrCauseLowPowerExit == RSTMGR_RESET_INFO_CAUSE_LOW_POWER_EXIT,
+    "kDifRstmgrCauseLowPowerExit must match the register definition!");
+_Static_assert(kDifRstmgrCauseWatchdog == RSTMGR_RESET_INFO_CAUSE_WATCHDOG,
+               "kDifRstmgrCauseWatchdog must match the register definition!");
+_Static_assert(
+    kDifRstmgrCauseSecurityEscalation == RSTMGR_RESET_INFO_CAUSE_ESCALATION,
+    "kDifRstmgrCauseSecurityEscalation must match the register definition!");
+_Static_assert(
+    kDifRstmgrCauseNonDebugModuleRequest ==
+        RSTMGR_RESET_INFO_CAUSE_NON_DEBUG_MODULE,
+    "kDifRstmgrCauseNonDebugModuleRequest must match the register definition!");
+
+_Static_assert(RSTMGR_SPI_DEVICE_REGEN_SPI_DEVICE_REGEN ==
+                   RSTMGR_USB_REGEN_USB_REGEN,
+               "SPI and USB regen bit offsets are not the same");
+_Static_assert(RSTMGR_RST_SPI_DEVICE_N_RST_SPI_DEVICE_N ==
+                   RSTMGR_RST_USB_N_RST_USB_N,
+               "SPI and USB reset bit offsets are not the same");
+
+#define RSTMGR_REGEN_BIT_OFFSET RSTMGR_SPI_DEVICE_REGEN_SPI_DEVICE_REGEN
+#define RSTMGR_RESET_BIT_OFFSET RSTMGR_RST_SPI_DEVICE_N_RST_SPI_DEVICE_N
+
+static const bitfield_field32_t kRstmgrCausesField = {
+    .mask = RSTMGR_RESET_INFO_CAUSE_MASK,
+    .index = RSTMGR_RESET_INFO_CAUSE_OFFSET,
+};
+
+// Regen registers are "write 1 to clear" = write 1 to lock.
+static const ptrdiff_t kRstmgrRegenOffsetsMap[kDifRstmgrPeripheralLast + 1] = {
+        [kDifRstmgrPeripheralSpi] = RSTMGR_SPI_DEVICE_REGEN_REG_OFFSET,
+        [kDifRstmgrPeripheralUsb] = RSTMGR_USB_REGEN_REG_OFFSET,
+};
+
+// Reset registers, when set to 0, peripheral is held in reset.
+static const ptrdiff_t kRstmgrResetOffsetsMap[kDifRstmgrPeripheralLast + 1] = {
```

Comment:
AIUI, you don't need to include the array extent on a global; C will just infer it. (I used to believe this was C++-only but it seems to work correctly in C, too.)

Reply:
Done.